service/storage_service.cc

```diff
@@ -1056,6 +1049,51 @@ class topology_coordinator {
         co_return std::pair{gen_uuid, std::move(gen_mutations)};
     }
 
+    // Broadcasts all mutations returned from `prepare_new_cdc_generation_data` except the last one.
+    // Each mutation is sent in separate raft command. It takes `group0_guard`, and if the number of mutations
+    // is greater than one, the guard is dropped, and a new one is created and returned, otherwise the old one
+    // will be returned. Commands are sent in parallel and unguarded (the guard used for sending the last mutation
+    // will guarantee that the term hasn't been changed). Returns the generation's UUID, guard and last mutation,
+    // which will be sent with additional topology data by the caller.
+    //
+    // With the default commitlog segment size, `mutation_size_threshold` will be 4 MB. In large clusters e.g.
+    // 100 nodes, 64 shards per node, 256 vnodes cdc generation data can reach the size of 30 MB, thus
+    // there will be no more than 8 commands.
+    //
+    // In a multi-DC cluster with 100ms latencies between DCs, this operation should take about 200ms since we
+    // send the commands concurrently, but even if the commands were replicated sequentially by Raft,
+    // it should take no more than 1.6s which is incomparably smaller than bootstrapping operation
+    // (bootstrapping is quick if there is no data in the cluster, but usually if one has 100 nodes they
+    // have tons of data, so indeed streaming/repair will take much longer (hours/days)).
+    future<std::tuple<utils::UUID, group0_guard, canonical_mutation>> prepare_and_broadcast_cdc_generation_data(
+            locator::token_metadata_ptr tmptr, group0_guard guard, std::optional<bootstrapping_info> binfo) {
+        auto [gen_uuid, gen_mutations] = co_await prepare_new_cdc_generation_data(tmptr, guard, binfo);
+
+        if (gen_mutations.empty()) {
+            on_internal_error(slogger, "cdc_generation_data: gen_mutations is empty");
+        }
+
+        std::vector<canonical_mutation> updates{gen_mutations.begin(), gen_mutations.end()};
+
+        if (updates.size() > 1) {
+            release_guard(std::move(guard));
+
+            co_await parallel_for_each(updates.begin(), std::prev(updates.end()), [this, gen_uuid = gen_uuid] (canonical_mutation& m) {
```

Comment:
Why can't we send all mutations including the last one? If term changes we will fail to write the topology mutation. We do not need cdc mutation for that.

Reply:
If we send the last mutation in the `write_mutation` command, we would use a total of `n + 1` commands instead of `n-1 + 1` (where n is the number of mutations), so it's better to send it in `topology_change` (we need to send it after all `write_mutations`) with some small metadata (`transition_state`, `new_cdc_generation_data_uuid`, etc.). There is no reason for excluding the last mutation here, other than the optimization of a number of raft commands sent.

Reply:
Not sure the optimization worth the complication, but lets at least amend the commit message and may be write a comment above the function that we want the happy path to issue only one raft command instead of two.